akka-projection-grpc/src/main/scala/akka/projection/grpc/replication/scaladsl/Replication.scala

```diff
@@ -0,0 +1,205 @@
+ *                             in a set passed to EventProducer.grpcServiceHandler to create a single gRPC endpoint
+ * @param createSingleServiceHandler If only replicating one Replicated Event Sourced Entity and not using
+ *                                   Akka Projection gRPC this endpoint factory can be used to get a partial function
+ *                                   that can be served/bound with an Akka HTTP server
+ * @param entityTypeKey Entity type key for looking up the
+ * @param entityRefFactory Shortcut for creating EntityRefs for the sharded Replicated Event Sourced entities for
+ *                         sending commands.
+ * @tparam Command The type of commands the Replicated Event Sourced Entity accepts
+ */
+@ApiMayChange
+final class Replication[Command] private (
+    val eventProducerService: EventProducerSource,
+    val createSingleServiceHandler: () => PartialFunction[HttpRequest, Future[HttpResponse]],
+    val entityTypeKey: EntityTypeKey[Command],
+    val entityRefFactory: String => EntityRef[Command])
+
+@ApiMayChange
+object Replication {
+
+  private val log = LoggerFactory.getLogger(classOf[Replication[_]])
+
+  private val filteredEvent = Future.successful(None)
+
+  /**
+   * Called to bootstrap the entity on each cluster node in each of the replicas.
+   *
+   * Important: Note that this does not publish the endpoint, additional steps are needed!
+   */
+  def grpcReplication[Command, Event, State](settings: ReplicationSettings[Command])(
+      replicatedBehaviorFactory: ReplicationContext => EventSourcedBehavior[Command, Event, State])(
+      implicit system: ActorSystem[_]): Replication[Command] = {
+    require(
+      system.settings.config.getString("akka.actor.provider") == "cluster",
+      "Replicated Event Sourcing over gRPC only possible together with Akka cluster (akka.actor.provider = cluster)")
+
+    val allReplicaIds = settings.otherReplicas.map(_.replicaId) + settings.selfReplicaId
+
+    // set up a publisher
+    val onlyLocalOriginTransformer = Transformation.empty.registerAsyncEnvelopeOrElseMapper(envelope =>
+      envelope.eventMetadata match {
+        case Some(meta: ReplicatedEventMetadata) =>
+          if (meta.originReplica == settings.selfReplicaId) Future.successful(envelope.eventOption)
+          else filteredEvent // Optimization: was replicated to this DC, don't pass the payload across the wire
+        case _ =>
+          throw new IllegalArgumentException(
+            s"Got an event without replication metadata, not supported (pid: ${envelope.persistenceId}, seq_nr: ${envelope.sequenceNr})")
+      })
+    val eps = EventProducerSource(
+      settings.entityTypeKey.name,
+      settings.streamId,
+      onlyLocalOriginTransformer,
+      settings.eventProducerSettings)
+
+    // sharding for hosting the entities and forwarding events
+    val replicatedEntity =
+      ReplicatedEntity(settings.selfReplicaId, Entity(settings.entityTypeKey) { entityContext =>
```

Comment:
There are rather many important settings of the `Entity` that we don't expose to the user. E.g. defining the stop message. It's nice that the user doesn't have to define the full ReplicatedEntity or Entity but we might need something to make it possible to amend the Entity settings?

Reply:
Started on this in 62e4248

I think that API is fine but I'll probably need to split up config impls somehow to actually implement it for both Java and Scala (no way to turn every type related to entities back and forth between Java to be able to turn a Java Entity => Entity into a Scala Entity => Entity).

Reply:
Split done in [1e0cbba](https://github.com/akka/akka-projection/pull/757/commits/1e0cbba3ac148d291271205b9bd9a9c496c7ef73)